lib/org/chromium/webidl/Diff.js

```diff
@@ -0,0 +1,247 @@
+// Copyright 2017 The Chromium Authors. ALl rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+'use strict';
+
+foam.CLASS({
+  package: 'org.chromium.webidl',
+  name: 'Diff',
+
+  requires: [
+    'org.chromium.webidl.DiffChunk',
+    'org.chromium.webidl.DiffStatus',
+  ],
+
+  documentation: `The diff algorithm that is performed on canonicalized
+    definitions.`,
+
+  properties: [
+    {
+      class: 'FObjectArray',
+      of: 'org.chromium.webidl.DiffChunk',
+      name: 'chunks',
+    },
+    {
+      name: 'excludedProps',
+      value: [ 'id', 'isCanonical', 'sources', 'source' ],
+    },
+  ],
+
+  methods: [
+    {
+      documentation: `diff expects to receive two maps of canonicalized
+        definitions. A diff will be performed for each definition, and an
+        array of DiffChunks will be returned for each difference found.`,
+      name: 'diff',
+      returns: 'Array',
+      code: function(left, right) {
+        this.chunks = [];
+
+        // Calculate union of leftKeys and rightKeys.
+        var leftKeys = Object.keys(left);
+        var rightKeys = Object.keys(right);
+
+        var keys = leftKeys.concat(rightKeys);
+        keys = keys.filter(function(key, index) {
+          return keys.indexOf(key) === index;
+        });
+
+        keys.forEach(function(key) {
+          // If we have an array of items (e.g. Enums / Typedef),
+          // these definitions should be the same.
+          // FUTURE: Handle multi definitions that are NOT the same.
+          if (Array.isArray(left[key])) this.dedupDefinition_(left, key);
+          if (Array.isArray(right[key])) this.dedupDefinition_(right, key);
+
+          // Perform diff with id object to track path and source.
+          var id = {
+            name: key,
+            path: '',
+            leftSources: left[key] ? left[key].sources : null,
+            rightSources: right[key] ? right[key].sources : null,
+          };
+
+          this.diff_(left[key], right[key], id);
+        }.bind(this));
+        return this.chunks;
```

Comment:
Can this component be made less stateful so it can be re-used by its runner? It looks like it performs its work synchronously, so, if we can't get rid of `changes` as a piece of local state, we can at least clone it, clear it, then return it.